src/cpp/dds/publisher/PublisherImpl.cpp

```diff
@@ -0,0 +1,492 @@
+    }
+
+    if (writer_qos.m_disablePositiveACKs.enabled &&
+        writer_qos.m_disablePositiveACKs.duration != c_TimeInfinite)
+    {
+        w_att.disable_positive_acks = true;
+        w_att.keep_duration = writer_qos.m_disablePositiveACKs.duration;
+    }
+
+    DataWriterImpl* impl = new DataWriterImpl(
+        this,
+        type_support,
+        topic_att,
+        w_att,
+        writer_qos,
+        att_.historyMemoryPolicy,
+        listener);
+
+    if(impl->writer_ == nullptr)
+    {
+        logError(PUBLISHER, "Problem creating associated Writer");
+        delete impl;
+        return nullptr;
+    }
+
+    DataWriter* writer = new DataWriter(impl);
+    impl->user_datawriter_ = writer;
+
+    //REGISTER THE WRITER
+    rtps_participant_->registerWriter(impl->writer_, topic_att, writer_qos);
+
+    {
+        std::lock_guard<std::mutex> lock(mtx_writers_);
+        writers_[topic_att.getTopicDataType().to_string()].push_back(impl);
+    }
+
+    return writer;
+}
+
+bool PublisherImpl::delete_datawriter(
+        DataWriter* writer)
+{
+    std::lock_guard<std::mutex> lock(mtx_writers_);
+    auto vit = writers_.find(writer->get_topic().getTopicName().to_string());
+    if (vit != writers_.end())
+    {
+        auto dw_it = std::find(vit->second.begin(), vit->second.end(), writer->impl_);
+        if (dw_it != vit->second.end())
+        {
+            vit->second.erase(dw_it);
+            return true;
+        }
+    }
+    return false;
+}
+
+DataWriter* PublisherImpl::lookup_datawriter(
+    const std::string& topic_name) const
+{
+    std::lock_guard<std::mutex> lock(mtx_writers_);
+    auto it = writers_.find(topic_name);
+    if (it != writers_.end() && it->second.size() > 0)
+    {
+        return it->second.front()->user_datawriter_;
+    }
+    return nullptr;
+}
+
+bool PublisherImpl::get_datawriters(
+        std::vector<DataWriter*>& writers) const
+{
+    std::lock_guard<std::mutex> lock(mtx_writers_);
+    for (auto vit : writers_)
+    {
+        for (DataWriterImpl* dw : vit.second)
+        {
+            writers.push_back(dw->user_datawriter_);
+        }
+    }
+    return true;
+}
+
+/* TODO
+bool PublisherImpl::suspend_publications()
+{
+    logError(PUBLISHER, "Operation not implemented");
+    return false;
+}
+*/
+
+/* TODO
+bool PublisherImpl::resume_publications()
+{
+    logError(PUBLISHER, "Operation not implemented");
+    return false;
+}
+*/
+
+/* TODO
+bool PublisherImpl::begin_coherent_changes()
+{
+    logError(PUBLISHER, "Operation not implemented");
+    return false;
+}
+*/
+
+/* TODO
+bool PublisherImpl::end_coherent_changes()
+{
+    logError(PUBLISHER, "Operation not implemented");
+    return false;
+}
+*/
+
+
+bool PublisherImpl::set_default_datawriter_qos(
+        const fastrtps::WriterQos& qos)
+{
+    if (&qos == &DATAWRITER_QOS_DEFAULT)
+    {
+        fastrtps::WriterQos def_qos;
+        default_datawriter_qos_.setQos(def_qos, true);
```

Comment:
In this case the function must return true
```suggestion
        default_datawriter_qos_.setQos(def_qos, true);
        return true
```